src/mavlink-router/endpoint.cpp

```diff
@@ -795,17 +843,21 @@ int UdpEndpoint::write_msg(const struct buffer *pbuf)
     };
 
     _stat.write.total++;
-    _stat.write.bytes += pbuf->len;
+    _stat.write.bytes += r;
 
-    /* Incomplete packet, we warn and discard the rest */
-    if (r != (ssize_t) pbuf->len) {
-        _incomplete_msgs++;
-        log_debug("Discarding packet, incomplete write %zd but len=%u", r, pbuf->len);
-    }
+    tx_buf.len = std::max(ssize_t(0), (ssize_t)tx_buf.len - r);
+    // memcpy isn't safe for overlapping regions
+    memmove(tx_buf.data, &tx_buf.data[r], tx_buf.len);
 
     log_debug("UDP: [%d] wrote %zd bytes", fd, r);
 
-    return r;
+    return tx_buf.len;
```

Comment:
I see nowhere where the pending timeout is cancelled in case the write happens due to buffer being to full -- see other comment regarding code structure recommendation

Reply:
There isn't API support in mavlink-router for cancelling timers at the epoll level. I'm not sure if it makes sense to add that, or if I should just ignore the wakeup if I didn't want the wakeup. Thoughts?

Reply:
if it is not possible to cancel then you would at least need to add code to ignore, so you would need to store "correct" deadline somewhere in order to distinguish planned/unplanned wakeups

not doing it means you will get early wakeups from dangling timers and will send prematurely, thus partly losing the effect of coalescing (maybe that is part of what is also observed in wireshark tests)

Reply:
From my understanding of [`timerfd_settime`](https://linux.die.net/man/2/timerfd_settime), it only tracks a single timeout for any given `fd`, and drops the old timeout if that `fd` had one registered. So we shouldn't have dangling wakeups. However, digging into mavlink-router it seems it configures the timer to wakeup every x milliseconds forever, not just one-shot. I'll disable the timer when it isn't needed.